akka-remote/src/main/scala/akka/remote/artery/tcp/ssl/TlsMagicSSLEngineProvider.scala

```diff
@@ -0,0 +1,70 @@
+package akka.remote.artery.tcp.ssl
+
+import java.security.SecureRandom
+import java.security.cert.X509Certificate
+import java.util.concurrent.atomic.AtomicReference
+
+import akka.actor.ActorSystem
+import akka.event.Logging
+import akka.event.MarkerLoggingAdapter
+import akka.remote.artery.tcp.SSLEngineProvider
+import akka.remote.artery.tcp.SecureRandomFactory
+import com.typesafe.config.Config
+import javax.net.ssl.SSLContext
+import javax.net.ssl.SSLEngine
+import javax.net.ssl.SSLSession
+
+import scala.concurrent.duration._
+
+// TODO: rename
+final class TlsMagicSSLEngineProvider(protected val config: Config, protected val log: MarkerLoggingAdapter)
+    extends SSLEngineProvider {
+
+  def this(system: ActorSystem) =
+    this(
+      system.settings.config
+      // TODO: when renaming, review the namespace in config
+        .getConfig("akka.remote.artery.ssl.tls-magic-engine"),
+      Logging.withMarker(system, classOf[TlsMagicSSLEngineProvider].getName))
+
+  // Cache support
+  private val SSLContextCacheTime: FiniteDuration =
+    config.getDuration("ssl-context-cache-ttl").toMillis.millis
+  private val contextRef = new AtomicReference[Option[Cache]](None)
+  private def getFactory: SslFactory = getCache.sslFactory
+  private def getCache: Cache = {
+    contextRef.get() match {
+      case Some(Cache(_, expired)) if expired.timeLeft =>
+        contextRef.get().get
+      case _ =>
+        // RNG support
+        // TODO: use "SecureRandomFactory.createSecureRandom(SSLRandomNumberGenerator, log)" instead,
+        //  I'm just trying to demonstrate usage of different implementations
+        val rng = new SecureRandom()
+        // This line is where all pieces are instantiated and put together.
+        val context = new SslFactory(config, new PemManagerProviders(config), rng)(log)
+        val cache = Cache(context, SSLContextCacheTime.fromNow)
```

Comment:
These are the 4 lines necessary (the `ConfigSSLEngineprovider` comment above).

The rest in this class is to implement caching, adapting interfaces, etc...

Note how this implementation caches a `Cache` instaned (I know, the name is horrible) which contains multiple fields I know I'll need later on (to crate Engine or verifySession).